Avoid reading JDBC connector table properties

kokosing opened the pull request:
Avoid reading JDBC connector table properties

This is achievied by Implementing
getTableBasicMetadata for JDBC connectors

ssheikin commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -330,20 +331,45 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+
+        return new ConnectorTableBasicMetadata(
+                getSchemaTableName(handle),
+                getColumnMetadata(session, handle).stream()
+                        .map(ColumnMetadata::toBasicMetadata)
```
Seems you can first `getColumnMetadata` and later  `.map(ColumnMetadata::toBasicMetadata)` only because `ColumnMetadata` does not have column properties. So actually it masks current inexistent problem with loading column properties for jdbc connectors, but problem would still exist if jdbc connector would have properties.

kokosing replied:
That is true. If we ever add column properties to JDBC connectors I don't believe that `JdbcColumnHandle` will be able to create `ColumnMetadata`. And surely we don't want `io.trino.plugin.jdbc.JdbcClient#getColumns` to query remote data source for column properties.

ssheikin commented on plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -83,19 +84,32 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
+                phoenixClient.getTableProperties(session, handle));
+    }
+
+    private List<ColumnMetadata> getColumnMetadata(ConnectorSession session, boolean rowkeyRequired, JdbcTableHandle handle)
```
order of parameters differs from https://github.com/trinodb/trino/pull/6975/commits/eaf66717f28023799cdc2552b882587ba0cb6eb5#diff-b2770586e716078fd17f177ce4cd0868628f704bc0cd81f3fb5f3d4def54e882L91

`rowkeyRequired` is always `false`.

kokosing commented:
This approach 4 from https://docs.google.com/document/d/1LqzlflWo4m6DySKihE1LSsDhqhOisAOET45pERJKEa4/edit#

kokosing commented:
AC

findepi commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -507,12 +507,7 @@ protected Scope visitRefreshMaterializedView(RefreshMaterializedView refreshMate
-            else {
-                analysis.setSkipMaterializedViewRefresh(false);
-            }
+            analysis.setSkipMaterializedViewRefresh(metadata.getMaterializedViewFreshness(session, name).isMaterializedViewFresh());
```
Why `targetTableHandle.isPresent()` is being removed?

kokosing replied:
It is already checked two lines above

findepi commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -2266,7 +2261,7 @@ private void analyzeWindowFunctions(QuerySpecification node, List<Expression> ou
                 // TODO get function requirements from window function metadata when we have it
                 String name = windowFunction.getName().toString().toLowerCase(ENGLISH);
-                if (name.equals("lag") || name.equals("lead")) {
+                if ("lag".equals(name) || "lead".equals(name)) {
```
We don't have convention either way, and the `name` cannot be null, so this change isn't necessary

findepi reviewed (commented):
"Cleanup in StatementAnalyzer"

findepi commented on core/trino-main/src/main/java/io/trino/metadata/Metadata.java:
```diff
@@ -113,6 +113,14 @@
 
+    /**
+     * Return basic metadata for the specified table handle.
+     * This is method should be preferred over {@link #getTableMetadata(Session, TableHandle)} as it is cheaper.
```
s/is//

findepi replied:
bump

findepi commented on core/trino-main/src/main/java/io/trino/metadata/TableBasicMetadata.java:
```diff
@@ -0,0 +1,71 @@
+    {
+        return metadata.getColumns().stream()
+                .filter(columnMetadata -> columnMetadata.getName().equals(name))
+                .findFirst()
```
`.collect(toOptional())` would throw on duplicates, which is more appropriate here.

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnMetadata.java:
```diff
@@ -135,6 +135,11 @@ public boolean isHidden()
 
+    public ColumnBasicMetadata toBasicMetadata()
+    {
+        return new ColumnBasicMetadata(name, type, nullable, hidden);
```
If the intention is that `ColumnBasicMetadata` is a subset of information, maybe `ColumnBasicMetadata` should be accepted in the `ColumnMetadata` ctor and be stored as a field? This could allow connectors to maintain mapping once.

However, I am not sure whether we can want to lock ourselves in one being subset of the other -- logically it make sense, but there can be a situation, where this isn't helpful. For this however, we would want this method (`toBasicMetadata()`) not to exist at all. (or exist as deprecated for the transition period, if needed)

thoughts?

kokosing replied:
> If the intention is that ColumnBasicMetadata is a subset of information, maybe ColumnBasicMetadata should be accepted in the ColumnMetadata ctor and be stored as a field?

Sounds good to me. That was my intention. I wanted to leave it optional for connectors to provide basic metadata. For example in case of Hive it does not really matter if we provide basic or non-basic as it translates to same number of HMS calls.

kokosing replied:
> be stored as a field

That was my plan too. Also @martint mentioned that is the way he would prefer. However it is not that simple as there is not only ConnectorTableSchema, but also ColumnSchema. So I would need to somehow split `ColumnMetadata` in `ConnectorTableMetadata`.

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java:
```diff
@@ -71,6 +72,15 @@ public SchemaTableName getTable()
+                table,
+                columns.stream()
+                        .map(ColumnMetadata::toBasicMetadata)
+                        .collect(Collectors.toUnmodifiableList()));
```
static import

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java:
```diff
@@ -0,0 +1,92 @@
+            return false;
+        }
+        ColumnBasicMetadata that = (ColumnBasicMetadata) o;
+        return nullable == that.nullable && hidden == that.hidden && name.equals(that.name) && type.equals(that.type);
```
put on separate lines

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnBasicMetadata.java:
```diff
@@ -0,0 +1,92 @@
+        sb.append(", type=").append(type);
+        sb.append(", nullable=").append(nullable);
+        sb.append(", hidden=").append(hidden);
+        sb.append('}');
```
My IDE generates toString using Joiner.

kokosing replied:
I wanted to be in line with `io.trino.spi.connector.ColumnMetadata#toString`. My IDE allows me to choose how to generate `toString`.

Is there any difference? If so then it would be nice to update other `toString` methods in SPI.

findepi reviewed (commented):
"Introduce ConnectorTableBasicMetadata"

findepi commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java:
```diff
@@ -108,6 +109,11 @@ public ColumnMetadata getColumnMetadata()
 
+    public ColumnBasicMetadata getColumnBasicMetadata()
+    {
+        return new ColumnBasicMetadata(columnName, columnType, nullable, false);
```
why not `getColumnMetadata(). toBasic()`?

kokosing replied:
I wanted to avoid intermediate object creation. Do you think, it is worth to create?

findepi commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -330,20 +331,37 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+                jdbcClient.getTableProperties(session, handle));
+    }
+
+    protected static SchemaTableName getSchemaTableName(JdbcTableHandle handle)
```
This method is extracted from a hack (see TODO inside) and hopefully can stay `private`, unless you already see a need to expose this.

kokosing replied:
> unless you already see a need to expose this.

I do, phoenix connector potentially can also have `!handle.isNamedRelation()`, but it was missing a TODO like this there.

Praveen2112 replied:
Maybe as a separate commit ?

findepi commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:
```diff
@@ -0,0 +1,78 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
```
let's document why
```
// Single-threaded because of shared mutable state, e.g. onGetTableProperties
```

findepi commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:
```diff
@@ -0,0 +1,78 @@
+    @Test
+    public void testGetTablePropertiesIsCalled()
+    {
+        AtomicBoolean getTablePropertiesIsCalled = new AtomicBoolean();
```
AtomicInteger to assess call count

findepi commented on plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -83,19 +84,32 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
+                phoenixClient.getTableProperties(session, handle));
+    }
+
+    private List<ColumnMetadata> getColumnMetadata(ConnectorSession session, JdbcTableHandle handle)
```
nit: or `Stream`

findepi reviewed (approved):
Alternative would be to move properties out of the current metadata objects, as attempted here https://github.com/trinodb/trino/pull/6966.
The sheer volume of changes can put off, but it's also partially already done.
However, it also means that change cannot be done in a backwards compatible manner, which is an obvious downside, and doesn't guarantee no similar problem in the future, where again CREATE TABLE needs a different set of information about a table than regular SELECT (e.g. indexes; from engine perspective).
For this reason, I  LGTM this PR, even though it adds duplication of table/column metadata classes.

findepi commented:
@martint @losipiuk @electrum you may be interested in taking a look

martint reviewed (changes requested):
The concept of basic vs non-basic metadata is confusing. What goes where? When does something qualify as a basic vs non-basic piece of metadata?

I would rather we split the API into two parts as in #6966.

For the purpose of gradual deprecation, we can add the new API along side the existing one, mark the "properties" part as deprecated for a while (the engine would call one, then the other, if the first one doesn't produce any results), migrate all connectors to the new API and leave some time buffer for third-party connectors to do so. Then we can get rid of the old API.

kokosing commented:
> I would rather we split the API into two parts as in #6966.

@martint I admit that #6966 sounds promising. But here is my concern:

Currently are two ways engine communicates with connector in regards to a table. By using table handle or using table metadata. These two objects match to each other, they are some kind equivalent to each other. We use handle for existing tables and we use metadata for not yet existing tables (create table, CTAS). We also use metadata for existing tables too, to simplify code paths so same code path is executed for CTAS and insert.

If table does not exists, we are forced to use table metadata. If we split it with #6966 then there will be no single class that would be a complete representation of a table, hence in all places where today metadata is used we would need to pass the pair.

The worst part is that someday someone would like to name that pair. Then the name could be `ConnectorTableFullMetadata`. Then we could end up here, but in a bit different shape.

> The concept of basic vs non-basic metadata is confusing. What goes where? When does something qualify as a basic vs non-basic piece of metadata?

I hope this is a naming issue. Rule of thumb, all goes to non-basic and only things that are needed to run non DDL queries goes to the basic.

findepi commented:
> The worst part is that someday someone would like to name that pair. Then the name could be ConnectorTableFullMetadata. Then we could end up here, but in a bit different shape.


Passing a pair as individual parameters -- ConnectorTableMetadata and table/column properties -- solves the problem we have today, but does not provide for future extendibility easily
The thing to consider is -- what if we introduce new type of information describing the table, which we cannot or do not want to put into our slimmed down ConnectorTableMetadata.

jirassimok commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:
```diff
@@ -161,6 +161,18 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
+     *
+     * @throws RuntimeException if table handle is no longer valid
+     */
+    default ConnectorTableBasicMetadata getTableBasicMetadata(ConnectorSession session, ConnectorTableHandle table)
```
How about calling it `getBasicTableMetadata`? No matter how I turn `getTableBasicMetadata` around in my head, I can't make it parse as anything like "get basic metadata for table."

The more direct `getBasicMetadataForTable` also seems like a good name to me.

I think even `getConnectorTableBasicMetadata` is better than `getTableBasicMetadata`, but it's very verbose, and `get[ReturnType]` is not a very useful function name (but I think it's better than the current version, which is neither the return type nor phrase with similar meaning).

kokosing replied:
Renamed it to `ConnectorTableSchema` after discussion with @martint .

kokosing reviewed (commented):
After a discussion with @martint we decided that this should be named `TableSchema`. We even prefered `TableSchemaDefinition` to avoid confusion with schemas (part of catalog and group of tables`, but `ConnectorTableSchemaDefinition` was too long to me.

`TableSchema` represents a schema of table. See `CREATE TABLE <table name> (<table schema>) WITH <table properties>`. However note that columns can also have properties.

Comments addressed, please take a look.

findepi commented on core/trino-main/src/main/java/io/trino/metadata/Metadata.java:
```diff
@@ -114,6 +114,16 @@
+     * Return table schema definition for the specified table handle.
+     * Table schema definition is a set of information
+     * required by semantic analyzer to analyze the query.
+     * This is method should be preferred over {@link #getTableMetadata(Session, TableHandle)} as it is cheaper.
```
Instead of (or additionally to) having this sentence here, add a plain `@see` pointing to `getTableMetadata`
In `getTableMetadata` add `@see` poitining here and recommending use of this method, as it returns less information

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:
```diff
@@ -161,6 +161,20 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
+    /**
+     * Return table schema definition for the specified table handle.
+     * Table schema definition is a set of information
+     * required by semantic analyzer to analyze the query.
```
Do we need to define the content of the TableSchema by referring to the engine internal?

connector implementor may not care what is semantic analyzer and what other analyzers we have internally.

if you want to refer to usecase, maybe it would be better to say that this is the information required when planning a SELECT query?

Note however, that even in such case, we call other methods, like getTableProperties

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:
```diff
@@ -161,6 +161,20 @@ default ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session
+     * Table schema definition is a set of information
+     * required by semantic analyzer to analyze the query.
+     * This method is useful when getting full table metadata is expensive.
+     * Engine will try to leverage basic table metadata whenever possible.
```
"basic table metadata" -- this doens't exist

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:
```diff
@@ -0,0 +1,95 @@
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class ColumnSchema
```
We may need a build for this, like we have for `ColumnMetadata`

as we strive for this class to be as small as possible, i don't believe we won't extend it in the future with additional information

for example we have nullable, we could add some other contraint information

findepi commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:
```diff
@@ -0,0 +1,95 @@
+        return type;
+    }
+
+    public boolean isNullable()
```
seems unused today. Curious about your thinking whether to include it here or not.

kokosing replied:
Planner might like to use it.

findepi replied:
I agree. Planner may want to use this and other attributes, this is why i suggested adding bulder (https://github.com/trinodb/trino/pull/6975#discussion_r587386504) for future compatibility.

Anyway, we may never need this. This information is already available via `ConnectorTableProperties` -> predicate -> `Domain`  -> contains null?

so let's remove it for now

findepi replied:
if you prefer to keep it, add a comment. otherwise someone may do "Remove unused field" tomorrow, pruning it :)

findepi commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcTableProperties.java:
```diff
@@ -0,0 +1,79 @@
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
+// Single-threaded because of shared mutable state, e.g. onGetTableProperties
```
comment should be before `@Test(singleThrea..`

kokosing commented:
AC.

@martint Please take a look.

kokosing commented:
@martint ping

kokosing commented:
CI hit: https://github.com/trinodb/trino/issues/7468

martint commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -483,7 +483,7 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     targetTable,
                     Optional.empty(),
                     Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
```
This doesn't seem to belong in this commit

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:
```diff
@@ -0,0 +1,138 @@
+
+        public ColumnSchema build()
+        {
+            return new ColumnSchema(
```
These can all go in the same line.

martint replied:
This is still not resolved. Did you forget to push your changes?

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder("ColumnBasicMetadata{");
```
s/sb/builder/

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public String toString()
+    {
+        final StringBuilder sb = new StringBuilder("ConnectorTableSchema{");
```
s/sb/builder

Also, remove `final`

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableMetadata.java:
```diff
@@ -78,7 +78,7 @@ public ConnectorTableSchema getTableSchema()
                 columns.stream()
                         .map(ColumnMetadata::getColumnSchema)
-                        .collect(Collectors.toUnmodifiableList()));
+                        .collect(toUnmodifiableList()));
```
Unrelated change

martint reviewed (commented):
Some minor comments. I only reviewed the engine parts.

kokosing commented:
CI hit: https://github.com/trinodb/trino/issues/4984

kokosing commented:
@martint rebased, and addressed comments.

Praveen2112 commented on core/trino-main/src/main/java/io/trino/metadata/TableSchema.java:
```diff
@@ -0,0 +1,72 @@
+    {
+        return tableSchema.getColumns().stream()
+                .filter(columnMetadata -> columnMetadata.getName().equals(name))
+                .collect(toOptional())
```
nit : How about `findFirst`. Won't it be a bit easy to read ?

Praveen2112 commented on core/trino-spi/src/main/java/io/trino/spi/connector/ColumnSchema.java:
```diff
@@ -129,10 +129,7 @@ public Builder setHidden(boolean hidden)
-                    name,
-                    type,
-                    hidden);
+            return new ColumnSchema(name, type, hidden);
```
This can be applied in the previous commit right ?

Praveen2112 commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableSchema.java:
```diff
@@ -44,10 +44,10 @@ public SchemaTableName getTable()
-        sb.append(", columns=").append(columns);
-        sb.append('}');
-        return sb.toString();
+        return new StringBuilder("ConnectorTableSchema{")
```
Can we move this previous commit ?

Praveen2112 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java:
```diff
@@ -108,6 +109,11 @@ public ColumnMetadata getColumnMetadata()
 
+    public ColumnSchema getColumnSchema()
+    {
+        return ColumnSchema.from(getColumnMetadata());
```
Any reason for not using `Builder` ?

Praveen2112 reviewed (approved):
Minor changes. LGTM...

ssheikin commented:
:tada:
